### PR TITLE
Create more efficient diffs

### LINKED ISSUE #5: Use diffs for proposer finalizer policy and proposer policy in block_header extension

### PR DESCRIPTION
**Pull request by heifner**
Use a more efficient diff for `finalizer_policy` and `proposer_policy`.

Resolves #5

### CODE REVIEW

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 1]
```diff
@@ -24,87 +34,138 @@ template <typename T, typename SizeType = size_t, template<typename Y, typename.
...
 
-   /// @param diff the diff_result created from diff(source, target), apply_diff(std::move(source), diff_result) => target
-   /// @param container the source of diff(source, target) to modify using the diff_result to produce original target
-   /// @return the modified container now equal to original target
+   /// @param diff the diff_result created from diff(orig, target), apply_diff(std::move(orig), diff_result) => target
+   /// @param container the orig of diff(orig, target) to modify using the diff_result to produce the target
+   /// @return the modified container now equal to target
    template <typename X>
    requires std::same_as<std::decay_t<X>, diff_result>
    static Container<T> apply_diff(Container<T>&& container, X&& diff) {
```
Comment (linh2931) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:142]: Should `container` be sorted first?

It is also better to be named `original` or `orig`.

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 2]
```diff
@@ -24,87 +34,138 @@ template <typename T, typename SizeType = size_t, template<typename Y, typename.
...
-      // Remove from the source based on diff.remove_indexes
-      std::ptrdiff_t offset = 0;
-      for (SizeType index : diff.remove_indexes) {
-         container.erase(container.begin() + index + offset);
-         --offset;
+      // Remove from the container based on diff.remove_indexes which is descending order
```
Comment (linh2931) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:143]: Another possible way to do apply_diff is:
1. Have a fresh target.
2. Skip every index in `diff.remove_indexes`
3. Insert every value in `diff.insert_indexes` into the target
4. Move value in diff.move_indexes to right location in target.

This way the erases/copies can be avoided, and seems less code.

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 3]
```diff
@@ -24,87 +34,138 @@ template <typename T, typename SizeType = size_t, template<typename Y, typename.
...
+            ++orig_index;
+            ++cur_index;
          }
       }
 
+      std::ranges::sort(result.remove_indexes, [](auto&& lhs, auto&& rhs) { return lhs > rhs; }); // descending
+      std::ranges::sort(result.insert_indexes, [](auto&& lhs, auto&& rhs) { return lhs.first < rhs.first; });
+      std::ranges::sort(result.move_indexes,   [](auto&& lhs, auto&& rhs) { return lhs.first < rhs.first; });
+
+      // update move indexes
```
Comment (linh2931) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:112]: This section of updating move indexes is quite complicated. Can it somehow be avoided by tweaking other parts of the algorithm?

File: libraries/libfc/include/fc/container/ordered_diff.hpp [hunk 4]
```diff
@@ -24,87 +34,138 @@ template <typename T, typename SizeType = size_t, template<typename Y, typename.
 requires std::equality_comparable<T> && std::random_access_iterator<typename Container<T>::iterator>
 class ordered_diff {
 public:
+
+   /// All indexes are in reference to the target. Apply in order: remove, insert, move.
```
Comment (linh2931) [inline libraries/libfc/include/fc/container/ordered_diff.hpp:38]: Isn't `remove_indexes` for original?

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: PERFORMANCE
summary: Implement a more efficient diff for finalizer_policy and proposer_policy.
Note:end

**Comment by heifner** [comment]
Not going to make release. May re-visit for a future consensus release.

**Pull request closed without merging**